### PR TITLE
[JIT] Throw error when no_grad is used as a decorator

### LINKED ISSUE #44768: [JIT] function decorated with @torch.no_grad() can not be exported when another function is called inside it.
## 🐛 Bug

A function (decorated with @torch.no_grad()) can not be exported to torchscript if another function is called inside it.

## To Reproduce
```
def g(x):
    return x + 1

@torch.no_grad()
def f(x):
    return g(x)

torch.jit.script(f)
```
and we will get:
```
RuntimeError: 
undefined value g:
  File "torch/autograd/grad_mode.py", line 13
@torch.no_grad()
def f(x):
    return g(x)
           ~ <--- HERE
```

## Environment
 - PyTorch Version (e.g., 1.0): master
 - OS (e.g., Linux): linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): build from source
 - Python version: 3.6
 - CUDA/cuDNN version: 10.0/7.6.5

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50129 [JIT] Throw error when no_grad is used as a decorator**

**Summary**
JIT supports `torch.jit.no_grad` as a context manager, but not as a
decorator. When it is used as the latter, it compiles successfully
without any errors, but there are no IR nodes emitted to implement
`no_grad` functionality.

This commit adds code in `get_jit_def` that throws an
`torch.jit.frontend.NotSupportedError` when a function decorated with
`@torch.no_grad` is encountered.

**Test Plan**
This commit adds a unit test to `test_with.py` (since that's where the
other `no_grad` tests are) to check that an error is thrown when
scripting functions decorated with `@torch.no_grad`.

**Fixes**
This commit fixes #44768.

Differential Revision: [D25855125](https://our.internmc.facebook.com/intern/diff/D25855125)

### CODE REVIEW

File: torch/jit/frontend.py [hunk 1]
```diff
@@ -217,6 +217,35 @@ def remove_prefix(text, prefix):
...
+            return f"{get_decorator_name(decorator.value)}.{decorator.attr}"
+        elif isinstance(decorator, ast.Name):
+            return decorator.id
+        else:
+            raise RuntimeError(f"Unsupported node type: {type(decorator)}")
+
+    for decorator in fn_def.decorator_list:
+        if isinstance(decorator, ast.Call):
+            decorator_name = get_decorator_name(decorator)
+            if decorator_name == "torch.no_grad":
```
Comment (gmagogsfm) [inline torch/jit/frontend.py:244]: Would this be safe against things like import aliasing?

import torch as to
from torch import no_grad

If it is not safe, could we fetch object according to name and compare against actual `torch.no_grad` object?

Comment (SplitInfinity) [reply torch/jit/frontend.py:244]: Good point, I'm so used to seeing `@torch.no_grad` everywhere that I didn't realize. Yes, I think it is possible to fetch the object and compare against `torch.no_grad`.

Comment (SplitInfinity) [reply torch/jit/frontend.py:244]: Done.

**Review by gmagogsfm — APPROVED** [review]
Looks good! Thanks for making this change.

I do have a high level question, though. `check_fn_decorators_supported` is phrased to check whether all decorators present are supported. And it is currently using a blocklist-based approach (only `torch.no_grad` is excluded). Does this mean all other decorators are supported? Should we consider an allowlist-based approach instead if that's not the case?

**Comment by SplitInfinity** [comment]
> Looks good! Thanks for making this change.
> 
> I do have a high level question, though. `check_fn_decorators_supported` is phrased to check whether all decorators present are supported. And it is currently using a blocklist-based approach (only `torch.no_grad` is excluded). Does this mean all other decorators are supported? Should we consider an allowlist-based approach instead if that's not the case?

A decorator itself can be user-defined or be part of a `torch`, and it's possible that a user's decorator is harmless as far as scripting is concerned. Except for `@ignore` and `@unused`, decorators in general are completely ignored by TorchScript, and Python will have "evaluated" (not sure if I'm using the right word here) them before scripting. `no_grad` is also special because it is used both as a context manager and a decorator; this is only true for `dual_level` and `enable_grad` (neither of which we support, so it might be worth adding them to this blocklist).

**Pull request closed without merging**